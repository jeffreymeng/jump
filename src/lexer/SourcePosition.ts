/**
 * A SourcePosition refers to a specific position in the source.
 * Each character in the source has a unique position.
 */
export default class SourcePosition {
	/**
	 * @param source - The text of the source code.
	 * @param index - The index of the character within the source text.
	 */
	constructor(protected readonly source: string, protected readonly index: number) {
		if (index >= source.length) {
			throw new Error("SourcePosition expected an index less than the length of source.")
		}
	}

	/**
	 * Returns the line of the index in the given source text. Lines are 1-indexed.
	 */
	public getLine(): number {
		return this.source.substring(0, this.index).split("\n").length
	}

	/**
	 * Returns the index of the character within its line. Columns are
	 * 1-indexed.
	 */
	public getCol(): number {
		// given the column, look backwards for the closest newline strictly
		// before the current index, since if the current character is a
		// newline that doesn't count.
		return this.index - this.source.lastIndexOf("\n", this.index - 1);
	}

}
