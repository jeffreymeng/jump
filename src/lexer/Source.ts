import SourcePosition from "./SourcePosition";
import { JumpSyntaxError } from "./Lexer";

export default class Source {
	protected _index = 0;
	public readonly text: string;

	/**
	 * Returns the current index of the source.
	 */
	public get index() {
		return this._index;
	}

	constructor(text: string) {
		this.text = text;
	}

	/**
	 * Returns the next character from the source, updating the line and position as necessary.
	 * If no next character exists, returns an empty string.
	 */
	public next(): string {
		if (this._index >= this.text.length) {
			throw new JumpSyntaxError(
				"Unexpected end of input.",
				new SourcePosition(this.text, this._index)
			);
		}
		const c = this.text[this._index];
		this._index++;
		return c;
	}

	/**
	 * Returns the next character from the source without changing the current position.
	 * @param offset - If provided, the ith character after the current one is returned.
	 * If `i` is out of range, an empty string will be returned.
	 */
	public peekNext(offset = 0): string {
		if (this._index + offset >= this.text.length) {
			return "";
		}

		return this.text[this._index + offset];
	}

	/**
	 * Returns the next `len` characters from the source without changing the current position.
	 * @param len - The number of characters to return, or `0` to return the rest of source. Must be in range
	 */
	public peekNextCharacters(len = 0): string {
		if (len === 0) {
			return this.text.substring(this._index);
		}

		if (this._index + len >= this.text.length) {
			throw new JumpSyntaxError(
				"Unexpected end of file",
				new SourcePosition(this.text, this._index)
			);
		}
		return this.text.substring(this._index, this._index + len);
	}

	/**
	 * Consumes characters until any character inside endChars is found.
	 * The end character is consumed, but only the characters from
	 * the current position (inclusive) to the last character before
	 * the end character are returned. The string must eventually contain
	 * one of the endChars.
	 * @param endChars - A string containing one or more characters that will cause consumeUntil()
	 *  to stop consuming characters.
	 * @param errorMessageIfNotFound - If the string does not eventually contain
	 * an endChar, then this message will be used instead of the default error message.
	 */
	public consumeUntil(
		endChars: string,
		errorMessageIfNotFound = "Unexpected end of input."
	): string {
		const consumed = this.consumeWhile((c) => !endChars.includes(c));
		// consume the character that caused the predicate to fail, unless
		// we reached the end of the string already (and thus nothing caused
		// this to throw), which would cause an error.
		if (this._index < this.text.length) {
			this.next();
		} else {
			throw new JumpSyntaxError(
				errorMessageIfNotFound,
				this.getPosition()
			);
		}
		return consumed;
	}

	/**
	 * Consumes and returns characters until a character not in `chars` is found.
	 * @param chars - A string containing one or more characters to be consumed continuously.
	 */
	public consumeAll(chars: string): string {
		return this.consumeWhile((c) => chars.includes(c));
	}

	/**
	 * Consumes characters until the predicate is no longer true.
	 * All characters from the starting position until the predicate fails will be consumed and returned.
	 * The character that failed the predicate will NOT be consumed.
	 * @param predicate - A function returning true if a character passes the predicate.
	 */
	public consumeWhile(predicate: (c: string) => boolean): string {
		let buffer = "";
		while (this.hasNext() && predicate(this.peekNext())) {
			buffer += this.next();
		}
		return buffer;
	}

	/**
	 * Get a SourcePosition object corresponding to the `i`th character of the source,
	 * or the current index if none is provided.
	 */
	public getPosition(i = this._index) {
		return new SourcePosition(this.text, i);
	}

	/**
	 * Returns true if the source has a next token (e.g. has not reached the end yet).
	 */
	public hasNext() {
		return this._index < this.text.length;
	}
}
