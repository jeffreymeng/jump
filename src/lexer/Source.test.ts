import { describe, test, expect } from "@jest/globals";

import Source from "./Source";
import SourcePosition from "./SourcePosition";

describe("Source", () => {
	test("next & peeking functions", () => {
		const s = new Source("abcdefghijk");
		expect(s.peekNext()).toBe("a");
		expect(s.next()).toBe("a");
		expect(s.next()).toBe("b");
		expect(s.next()).toBe("c");
		expect(s.peekNext()).toBe("d");
		expect(s.peekNext()).toBe("d");
		expect(s.next()).toBe("d");
		expect(s.peekNextCharacters(3)).toBe("efg");
		expect(s.peekNextCharacters(4)).toBe("efgh");
		expect(() => s.peekNextCharacters(10)).toThrowErrorMatchingSnapshot();
		expect(s.peekNextCharacters()).toBe("efghijk");
		expect(() => s.peekNextCharacters("efghijk".length + 1)).toThrowErrorMatchingSnapshot();
	});

	test("peekNext returns empty string once source is over", () => {
		const s = new Source("a");
		expect(s.peekNext()).toBe("a");
		s.next();
		expect(s.peekNext()).toBe("");
	});

	test("getPosition", () => {
		const s = new Source(`aaaabb333`);
		s.consumeAll("a");
		expect(s.getPosition()).toStrictEqual(
			new SourcePosition("aaaabb333", 4)
		);
		s.consumeAll("b");
		expect(s.getPosition()).toStrictEqual(
			new SourcePosition("aaaabb333", 6)
		);
		s.consumeAll("3");
		expect(s.getPosition()).toStrictEqual(
			new SourcePosition("aaaabb333", 9)
		);
	});

	test("consumeAll", () => {
		const s = new Source(`aaaabb333`);
		expect(s.consumeAll("d")).toBe("");
		expect(s.consumeAll("ab")).toBe("aaaabb");
		expect(s.peekNextCharacters()).toBe("333");
		expect(s.consumeAll("3")).toBe("333");
		expect(s.index).toBe(9);
		expect(() => s.next()).toThrowErrorMatchingSnapshot();
	});

	test("consumeUntil", () => {
		const s = new Source(`aaaabb333`);
		expect(s.consumeUntil("b3")).toBe("aaaa");

		const s2 = new Source(`aaaabb344`);
		expect(s2.consumeUntil("3")).toBe("aaaabb");
		// it should have also consumed a 3
		expect(s2.peekNextCharacters()).toBe("44");
		expect(s2.index).toBe(s2.text.length - 2);
		expect(() => s2.consumeUntil("a")).toThrowErrorMatchingSnapshot();
	});

	test("consumeWhile", () => {
		const s = new Source(`aaaabb333`);
		expect(s.consumeWhile((c) => c === "a" || c === "b")).toBe("aaaabb");
		expect(s.next()).toBe("3");

		const s2 = new Source(`aaaabb333`);
		expect(s2.consumeWhile((c) => c !== "3")).toBe("aaaabb");
		expect(s2.peekNextCharacters()).toBe("333");
		expect(s2.consumeWhile(() => true)).toBe("333");
		expect(s2.index).toBe(s.text.length);
		expect(() => s2.next()).toThrowErrorMatchingSnapshot();
	});
});
