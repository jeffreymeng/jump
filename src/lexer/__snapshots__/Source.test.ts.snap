// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Source consumeAll 1`] = `"Unexpected end of input."`;

exports[`Source consumeUntil 1`] = `"Unexpected end of input."`;

exports[`Source consumeWhile 1`] = `"Unexpected end of input."`;

exports[`Source next & peeking functions 1`] = `"Unexpected end of file"`;

exports[`Source next & peeking functions 2`] = `"Unexpected end of file"`;
